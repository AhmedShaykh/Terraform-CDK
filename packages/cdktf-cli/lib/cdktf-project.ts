import { AbortController } from "node-abort-controller"; // polyfill until we update to node 16
import { SynthesizedStack, SynthStack } from "./synth-stack";
import { printAnnotations } from "./synth";
import { CdktfStack, StackUpdate } from "./cdktf-stack";
import { Errors } from "./errors";

export type ProjectUpdate =
  | {
      type: "synthesizing";
    }
  | {
      type: "synthesized";
      stacks: SynthesizedStack[];
      errorMessage?: string;
    }
  | StackUpdate;

function getSingleStack(
  stacks: SynthesizedStack[],
  stackName?: string,
  targetAction?: string
) {
  if (!stacks) {
    throw Errors.Internal(
      "Trying to access a stack before it has been synthesized"
    );
  }

  if (stackName) {
    const stack = stacks.find((s) => s.name === stackName);
    if (!stack) {
      throw Errors.Usage("Could not find stack: " + stackName);
    }

    return stack;
  }

  if (stacks.length === 1) {
    return stacks[0];
  }

  throw Errors.Usage(
    `Found more than one stack, please specify a target stack. Run cdktf ${
      targetAction || "<verb>"
    } <stack> with one of these stacks: ${stacks
      .map((s) => s.name)
      .join(", ")} `
  );
}

function getMultipleStacks(
  stacks: SynthesizedStack[],
  stackNames?: string[],
  targetAction?: string
) {
  if (!stackNames || !stackNames.length) {
    if (stacks.length === 1) {
      return [stacks[0]];
    }
    throw Errors.Usage(
      `Found more than one stack, please specify a target stack. Run cdktf ${
        targetAction || "<verb>"
      } <stack> with one of these stacks: ${stacks
        .map((s) => s.name)
        .join(", ")} `
    );
  }

  return stackNames.map((stackName) => getSingleStack(stacks, stackName));
}

// Returns the first stack that has no unmet dependencies
// An unmet dependency is a dependency that has not been deployed yet
function getStackWithNoUnmetDependencies(
  stackExecutors: CdktfStack[]
): CdktfStack | undefined {
  return stackExecutors
    .filter((executor) => executor.currentState !== "done")
    .find((executor) =>
      executor.stack.dependencies.every(
        (dependency) =>
          stackExecutors.find((executor) => executor.stack.name === dependency)
            ?.currentState === "done"
      )
    );
}

// Returns the first stack that has no dependents that need to be destroyed first
function getStackWithNoUnmetDependants(
  stackExecutors: CdktfStack[]
): CdktfStack | undefined {
  return stackExecutors
    .filter((executor) => executor.currentState !== "done")
    .find((executor) => {
      const dependantStacks = stackExecutors.filter((innerExecutor) =>
        innerExecutor.stack.dependencies.includes(executor.stack.name)
      );

      return dependantStacks.every((stack) => stack.currentState === "done");
    });
}

// Throws an error if there is a dependant stack that is not included
function checkIfAllDependantsAreIncluded(
  stacksToRun: SynthesizedStack[],
  allStacks: SynthesizedStack[]
) {
  const allDependants = new Set<string>();
  stacksToRun
    .map((stack) =>
      allStacks.filter((s) => s.dependencies.includes(stack.name))
    )
    .flat()
    .forEach((dependant) => allDependants.add(dependant.name));

  const stackNames = stacksToRun.map((stack) => stack.name);
  const missingDependants = [...allDependants].filter(
    (dependant) => !stackNames.includes(dependant)
  );

  if (missingDependants.length > 0) {
    throw Errors.Usage(
      `The following dependant stacks are not included in the stacks to run: ${missingDependants.join(
        ", "
      )}.`
    );
  }
}

// Throws an error if there is a dependency that is not included
// Cycles are detected on dependency creation at synthesis time
// Running this prevents us from being in a situation where we have to wait for a stack to be deployed
// that is not included to be run
function checkIfAllDependenciesAreIncluded(stacksToRun: SynthesizedStack[]) {
  const allDependencies = new Set<string>();
  stacksToRun
    .map((stack) => stack.dependencies)
    .flat()
    .forEach((dependency) => allDependencies.add(dependency));

  const stackNames = stacksToRun.map((stack) => stack.name);
  const missingDependencies = [...allDependencies].filter(
    (dependency) => !stackNames.includes(dependency)
  );

  if (missingDependencies.length > 0) {
    throw Errors.Usage(
      `The following dependencies are not included in the stacks to run: ${missingDependencies.join(
        ", "
      )}.`
    );
  }
}

export type SingleStackOptions = {
  stackName?: string;
};

export type MultipleStackOptions = {
  stackNames?: string[];
};

export type ExecutionOptions = MultipleStackOptions & {
  autoApprove?: boolean;
  ignoreMissingStackDependencies?: boolean;
};

export class CdktfProject {
  public stacks?: SynthesizedStack[];
  public hardAbort: () => void;

  private synthCommand: string;
  private outDir: string;
  private workingDirectory: string;
  private onUpdate: (update: ProjectUpdate) => void;
  private onLog?: (log: { stackName: string; message: string }) => void;
  private abortSignal: AbortSignal;

  constructor({
    synthCommand,
    outDir,
    onUpdate,
    onLog,
    workingDirectory = process.cwd(),
  }: {
    synthCommand: string;
    outDir: string;
    onUpdate: (update: ProjectUpdate) => void;
    onLog?: (log: { stackName: string; message: string }) => void;
    workingDirectory?: string;
  }) {
    this.synthCommand = synthCommand;
    this.outDir = outDir;
    this.workingDirectory = workingDirectory;
    this.onUpdate = onUpdate;
    this.onLog = onLog;
    const ac = new AbortController();
    this.abortSignal = ac.signal;

    this.hardAbort = ac.abort.bind(ac);
  }

  public getStackExecutor(
    stack: SynthesizedStack,
    opts: ExecutionOptions = {}
  ) {
    return new CdktfStack({
      ...opts,
      stack,
      onUpdate: this.onUpdate,
      onLog: this.onLog,
      abortSignal: this.abortSignal,
    });
  }

  public async synth() {
    this.onUpdate({
      type: "synthesizing",
    });
    const stacks = await SynthStack.synth(
      this.abortSignal,
      this.synthCommand,
      this.outDir,
      this.workingDirectory
    );

    printAnnotations(stacks);

    this.onUpdate({
      type: "synthesized",
      stacks,
    });

    return stacks;
  }

  public async diff(opts?: SingleStackOptions) {
    const stacks = await this.synth();
    const stack = this.getStackExecutor(
      getSingleStack(stacks, opts?.stackName, "diff")
    );
    await stack.diff();
    return stack.currentPlan;
  }

  public async deploy(opts: ExecutionOptions = {}) {
    const stacks = await this.synth();
    const stacksToRun = getMultipleStacks(stacks, opts.stackNames, "deploy");
    if (!opts.ignoreMissingStackDependencies) {
      checkIfAllDependenciesAreIncluded(stacksToRun);
    }

    const stackExecutors = stacksToRun.map((stack) =>
      this.getStackExecutor(stack, opts)
    );
    const next = opts.ignoreMissingStackDependencies
      ? () => stackExecutors.filter((stack) => stack.currentState !== "done")[0]
      : () => getStackWithNoUnmetDependencies(stackExecutors);

    let nextRunningExecutor = next();

    while (nextRunningExecutor) {
      await nextRunningExecutor.deploy();
      nextRunningExecutor = next();
    }

    const unprocessedStacks = stackExecutors.filter(
      (executor) => executor.currentState !== "done"
    );
    if (unprocessedStacks.length > 0) {
      throw Errors.External(
        `Some stacks failed to deploy: ${unprocessedStacks
          .map((s) => s.stackName)
          .join(", ")}. Please check the logs for more information.`
      );
    }
  }

  public async destroy(opts: ExecutionOptions = {}) {
    const stacks = await this.synth();
    const stacksToRun = getMultipleStacks(stacks, opts.stackNames, "destroy");

    if (!opts.ignoreMissingStackDependencies) {
      checkIfAllDependantsAreIncluded(stacksToRun, stacks);
    }
    const stackExecutors = stacksToRun.map((stack) =>
      this.getStackExecutor(stack, opts)
    );
    const next = opts.ignoreMissingStackDependencies
      ? () => stackExecutors.filter((stack) => stack.currentState !== "done")[0]
      : () => getStackWithNoUnmetDependants(stackExecutors);

    let nextRunningExecutor = next();

    while (nextRunningExecutor) {
      await nextRunningExecutor.destroy();
      nextRunningExecutor = next();
    }

    const unprocessedStacks = stackExecutors.filter(
      (executor) => executor.currentState !== "done"
    );
    if (unprocessedStacks.length > 0) {
      throw Errors.External(
        `Some stacks failed to destroy: ${unprocessedStacks
          .map((s) => s.stackName)
          .join(", ")}. Please check the logs for more information.`
      );
    }
  }

  public async fetchOutputs(opts?: SingleStackOptions) {
    const stacks = await this.synth();
    const stack = this.getStackExecutor(
      getSingleStack(stacks, opts?.stackName, "output")
    );
    return await stack.fetchOutputs();
  }
}
