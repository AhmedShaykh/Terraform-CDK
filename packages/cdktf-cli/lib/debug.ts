import * as path from "path";
import * as os from "os";
import { logger } from "./logging";
import { exec } from "./util";
import { terraformVersion } from "./terraform";
import { DISPLAY_VERSION } from "./version";

export function getLanguage(projectPath = process.cwd()): string | undefined {
  try {
    const cdktfJson = require(path.resolve(projectPath, "cdktf.json"));
    return cdktfJson.language;
  } catch {
    // We can not detect the language
    logger.debug(`Unable to detect language in ${projectPath}`);
    return undefined;
  }
}

async function getBinaryVersion(
  binary: string,
  versionCommand: string
): Promise<string | undefined> {
  try {
    const result = await exec(binary, [versionCommand], { env: process.env });
    return result.trim().replace(/\r?\n|\r/g, "");
  } catch (e) {
    logger.info(`Failed to run ${binary} ${versionCommand}: ${e}`);
    return undefined;
  }
}

export function getDotnetVersion() {
  return getBinaryVersion("dotnet", "--version");
}

export function getJavaVersion() {
  return getBinaryVersion("java", "--version");
}
export function getMavenVersion() {
  return getBinaryVersion("mvn", "--version");
}

export async function getPythonVersion() {
  return (
    (await getBinaryVersion("python3", "--version")) ||
    (await getBinaryVersion("python", "--version"))
  );
}
export function getPipVersion() {
  return getBinaryVersion("pip", "--version");
}
export function getPipenvVersion() {
  return getBinaryVersion("pipenv", "--version");
}

export function getGoVersion() {
  return getBinaryVersion("go", "version");
}

export function getNodeVersion() {
  return getBinaryVersion("node", "--version");
}

async function getNodeModuleVersion(
  packageName: string
): Promise<string | undefined> {
  let output;
  try {
    output = await exec("npm", ["list", packageName, "--json"], {
      env: process.env,
    });
  } catch (e) {
    logger.info(`Unable to run 'npm list ${packageName} --json': ${e}`);
    return undefined;
  }

  let json;
  try {
    json = JSON.parse(output);
  } catch (e) {
    logger.info(
      `Unable to parse output of 'npm list ${packageName} --json': ${e}`
    );
    return undefined;
  }

  if (
    !json.dependencies ||
    !json.dependencies[packageName] ||
    !json.dependencies[packageName].version
  ) {
    logger.info(
      `Unable to find '${packageName}' in 'npm list ${packageName} --json': ${output}`
    );
    return undefined;
  }

  return json.dependencies[packageName].version;
}

async function getPipenvPackageVersion(
  packageName: string
): Promise<string | undefined> {
  let output;
  try {
    output = await exec("pipenv", ["run", "pip", "show", packageName], {
      env: process.env,
    });
  } catch (e) {
    logger.info(`Unable to run 'pipenv run pip show ${packageName}': ${e}`);
  }

  // If we couldn't get the output using pipenv, try to get it using pip directly
  if (!output) {
    try {
      output = await exec("pip", ["show", packageName], {
        env: process.env,
      });
    } catch (e) {
      logger.info(`Unable to run 'pip show ${packageName}': ${e}`);
    }
  }

  if (!output) {
    return undefined;
  }

  const versionInfo = output
    .split("\n")
    .find((line) => line.startsWith("Version:"));

  if (!versionInfo) {
    logger.info(
      `Unable to find version in output of 'pipenv run pip show ${packageName}' / 'pip show ${packageName}': ${output}`
    );
    return undefined;
  }

  const version = versionInfo.split(":")[1].trim();
  return version;
}

export async function getPackageVersion(
  language: string,
  packageName: string
): Promise<string | undefined> {
  const noOp = async () => undefined; // eslint-disable-line @typescript-eslint/no-empty-function
  const getLibraryVersionMap: Record<
    string,
    (name: string) => Promise<string | undefined>
  > = {
    typescript: getNodeModuleVersion,
    python: getPipenvPackageVersion,
    go: noOp,
    csharp: noOp,
    java: noOp,
  };

  const libVersion = await (getLibraryVersionMap[language] || noOp)(
    packageName
  );
  return libVersion ?? undefined;
}

export async function collectDebugInformation() {
  const debugOutput: Record<string, string | null> = {};
  const language = getLanguage();
  debugOutput["language"] = language ?? null;
  debugOutput["cdktf-cli"] = DISPLAY_VERSION;
  debugOutput["node"] = (await getNodeVersion()) ?? null;
  if (language) {
    debugOutput["cdktf"] = (await getPackageVersion(language, "cdktf")) ?? null;
    debugOutput["constructs"] =
      (await getPackageVersion(language, "constructs")) ?? null;
    debugOutput["jsii"] = (await getPackageVersion(language, "jsii")) ?? null;
  }
  debugOutput["terraform"] = await terraformVersion;
  debugOutput["arch"] = os.arch();
  debugOutput["os"] = `${os.platform()} ${os.release()}`;

  switch (language) {
    case "python":
      debugOutput["python"] = (await getPythonVersion()) ?? null;
      debugOutput["pip"] = (await getPipVersion()) ?? null;
      debugOutput["pipenv"] = (await getPipenvVersion()) ?? null;
      break;
    case "java":
      debugOutput["java"] = (await getJavaVersion()) ?? null;
      debugOutput["maven"] = (await getMavenVersion()) ?? null;
      break;
    case "csharp":
      debugOutput["dotnet"] = (await getDotnetVersion()) ?? null;
      break;
    case "go":
      debugOutput["go"] = (await getGoVersion()) ?? null;
      break;
  }
  return debugOutput;
}
