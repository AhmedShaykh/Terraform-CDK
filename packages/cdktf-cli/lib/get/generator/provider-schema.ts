import * as fs from 'fs-extra';
import * as path from 'path';
import { promisify } from 'util';
import { exec, withTempDir } from '../../util';

const writeFile = promisify(fs.writeFile);
const terraformBinaryName = process.env.TERRAFORM_BINARY_NAME || 'terraform'

export interface ProviderSchema {
  format_version: '1.0';
  provider_schemas?: { [type: string]: Provider };
}

export interface Provider {
  provider: Schema;
  resource_schemas: { [type: string]: Schema };
  data_source_schemas: { [type: string]: Schema };
}

export interface Schema {
  version: number;
  block: Block;
}

export interface Attribute {
  type: AttributeType;
  description?: string;
  required?: boolean;
  optional?: boolean;
  computed?: boolean;
  sensitive?: boolean;
}

export type AttributeType =
  'string'
  | 'bool'
  | 'number'
  | [ 'set', AttributeType ]
  | [ 'map', AttributeType ]
  | [ 'list', AttributeType ]
  | [ 'object', { [attribute: string]: AttributeType } ];

export type BlockType = {
  nesting_mode: 'single' | 'map';
  block: Block;
} | {
  nesting_mode: 'list' | 'set';
  block: Block;
  min_items?: number;
  max_items?: number;
}

export interface Block {
  attributes: { [name: string]: Attribute };
  block_types: { [name: string]: BlockType };
}

export async function readSchema(providers: string[]): Promise<ProviderSchema> {
  const provider: { [name: string]: {} } = {};
  const requiredProviders: { [name: string]: { source?: string;  version?: string } } = { };

  for (const p of providers) {
    const [ fqname, version ] = p.split('@');
    const name = fqname.split('/').pop()
    if (!name) { throw new Error(`Provider name should be properly set in ${p}`) }

    provider[name] = {};
    requiredProviders[name] = { version, source: fqname };
  }
  let schema = '';

  await withTempDir('fetchSchema', async () => {
    const outdir = process.cwd();
    const filePath = path.join(outdir, 'providers.tf.json');
    // eslint-disable-next-line @typescript-eslint/camelcase
    await writeFile(filePath, JSON.stringify({ provider, terraform: { required_providers: requiredProviders }}));

    // todo: when implementing logging, we need to make sure we can show the terraform init
    // output if the log level is set to debug
    await exec(terraformBinaryName, [ 'init' ], { cwd: outdir });
    schema = await exec(terraformBinaryName, ['providers', 'schema', '-json'], { cwd: outdir });
    fs.unlinkSync(filePath)
  })

  return JSON.parse(schema);
}
