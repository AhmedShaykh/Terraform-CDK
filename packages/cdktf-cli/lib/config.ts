import * as fs from 'fs-extra';
import * as path from 'path';
import { Language } from './get/base';

const CONFIG_FILE = 'cdktf.json'
const DEFAULTS = {
  output: 'cdktf.out',
  codeMakerOutput: '.gen'
}

function isPresent(input: string[] | undefined): boolean {
  return Array.isArray(input) && input.length > 0
}

export interface Config {
  readonly app?: string;
  readonly language?: Language;
  readonly output: string;
  readonly codeMakerOutput: string;
  readonly terraformProviders?: string[];
  readonly terraformModules?: string[];
  checkCodeMakerOutput?: boolean;
}

export function readConfigSync(): Config {
  const configFile = path.join(process.cwd(), CONFIG_FILE)
  let config: Config = DEFAULTS;
  if (fs.existsSync(configFile)) {
    config = {
      ...config,
      ...JSON.parse(fs.readFileSync(configFile).toString())
    };
  }

  config.checkCodeMakerOutput = isPresent(config.terraformModules) || isPresent(config.terraformProviders)

  return config;
}