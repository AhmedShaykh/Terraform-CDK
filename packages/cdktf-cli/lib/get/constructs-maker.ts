import * as fs from 'fs-extra';
import * as path from 'path';
import { CodeMaker } from 'codemaker';
import { mkdtemp } from '../util';
import * as srcmak from 'jsii-srcmak';
import { TerraformModuleConstraint, TerraformDependencyConstraint } from '../config'
import { ProviderSchema, readSchema } from './generator/provider-schema';
import { TerraformProviderGenerator } from './generator/provider-generator';
import { ModuleGenerator } from './generator/module-generator';
import { ModuleSchema } from './generator/module-schema';
import { versionNumber } from '../../bin/cmds/version-check';
import { ReportParams, ReportRequest } from '../checkpoint'

const VERSION = versionNumber();

export enum Language {
  TYPESCRIPT = 'typescript',
  PYTHON = 'python',
  CSHARP = 'csharp',
  JAVA = 'java',
  GO = 'go',
}

export const LANGUAGES = [ Language.TYPESCRIPT, Language.PYTHON, Language.JAVA, Language.CSHARP, Language.GO ];

export interface GetOptions {
  readonly targetLanguage: Language;
  readonly codeMakerOutput: string;
  /**
   * Path to copy the output .jsii file.
   * @default - jsii file is not emitted
   */
  readonly outputJsii?: string;
}

export abstract class ConstructsMakerTarget {
  public readonly fileName: string;

  constructor(public readonly constraint: TerraformDependencyConstraint, public readonly targetLanguage: Language) {
    if (this.constraint instanceof TerraformModuleConstraint) {
      this.fileName = `${this.typesPath(this.constraint.fqn)}.ts`
    } else {
      this.fileName = `${this.typesPath(this.constraint.name)}.ts`
    }
  }

  public static from(constraint: TerraformDependencyConstraint, targetLanguage: Language) {
    if (constraint instanceof TerraformModuleConstraint) {
      return new ConstructsMakerModuleTarget(constraint, targetLanguage)
    } else {
      return new ConstructsMakerProviderTarget(constraint, targetLanguage)
    }
  }

  public get version() {
    return this.constraint.version
  }

  public get source() {
    return this.constraint.source
  }

  public get name() {
    return this.constraint.name
  }

  public get fqn() {
    return this.constraint.fqn
  }

  public get moduleKey() {
    return this.fqn.replace(/\//gi, '_')
  }

  public abstract get srcMakName(): string;
  public abstract get isModule(): boolean;
  public abstract get isProvider(): boolean;
  public abstract get trackingPayload(): Record<string, any>;
  protected abstract get simplifiedName(): string;


  protected abstract typesPath(name: string): string;
}

export class ConstructsMakerModuleTarget extends ConstructsMakerTarget {
  public spec?: ModuleSchema

  public get isModule() {
    return true
  }

  public get isProvider() {
    return false
  }

  public get srcMakName(): string {
    switch (this.targetLanguage) {
      case Language.JAVA, Language.CSHARP, Language.PYTHON, Language.GO:
        return this.simplifiedName;
      default:
        return this.constraint.fqn;
    }
  }

  public get trackingPayload() {
    return {
      name: this.name,
      fullName: this.source,
      version: this.version,
      type: 'module'
    }
  }

  protected typesPath(name: string): string {
    return `modules/${name}`;
  }

  protected get simplifiedName(): string {
    return this.fqn.replace(/\//gi, '.').replace(/-/gi, '_');
  }
}

export class ConstructsMakerProviderTarget extends ConstructsMakerTarget {
  public spec?: ProviderSchema

  public get isModule() {
    return false
  }

  public get isProvider() {
    return true
  }

  public get srcMakName(): string {
    switch (this.targetLanguage) {
      case Language.JAVA:
        // "null" is a reserved keyword and can't be used as a package name
        return this.isNullProvider ? "nullprovider" : this.simplifiedName;
      case Language.CSHARP:
        // "null" is a reserved keyword and can't be used as a namespace
        return this.isNullProvider ? "Providers.Null" : this.simplifiedName;
      case Language.PYTHON:
        return this.simplifiedName;
      case Language.GO:
        return this.simplifiedName;
      default:
        return this.constraint.fqn;
    }
  }

  public get trackingPayload() {
    return {
      name: this.name,
      fullName: this.source,
      version: this.version,
      type: 'provider'
    }
  }

  protected typesPath(name: string): string {
    return `providers/${name}/index`;
  }

  private get isNullProvider() {
    return this.constraint.name === "null"
  }

  protected get simplifiedName(): string {
    return this.name.replace(/\//gi, '.').replace(/-/gi, '_');
  }
}


export class ConstructsMaker {
  private readonly codeMakerOutdir: string;
  private readonly code: CodeMaker;
  private readonly targets: ConstructsMakerTarget[];

  constructor(private readonly options: GetOptions, private readonly constraints: TerraformDependencyConstraint[]) {
    this.codeMakerOutdir = path.resolve(this.options.codeMakerOutput);
    fs.mkdirpSync(this.codeMakerOutdir);
    this.code = new CodeMaker();
    this.targets = this.constraints.map(constraint => ConstructsMakerTarget.from(constraint, this.options.targetLanguage))
  }

  private async generateTypeScript() {
    const schema = await readSchema(this.targets);

    const moduleTargets: ConstructsMakerModuleTarget[] = this.targets.filter(target => target instanceof ConstructsMakerModuleTarget) as ConstructsMakerModuleTarget[]
    for (const target of moduleTargets) {
      target.spec = schema.moduleSchema[target.moduleKey]
    }

    const providerTargets: ConstructsMakerProviderTarget[] = this.targets.filter(target => target instanceof ConstructsMakerProviderTarget) as ConstructsMakerProviderTarget[];

    if (providerTargets.length > 0) {
      new TerraformProviderGenerator(this.code, schema.providerSchema, providerTargets);
    }

    if (moduleTargets.length > 0) {
      new ModuleGenerator(this.code, moduleTargets);
    }
  }

  public async generate() {
    await this.generateTypeScript();

    if (this.isJavascriptTarget) {
      await this.save()
    }

    if (!this.isJavascriptTarget || this.options.outputJsii) {
      for (const target of this.targets) {
        await mkdtemp(async staging => {
          // this is not typescript, so we generate in a staging directory and
          // use jsii-srcmak to compile and extract the language-specific source
          // into our project.
          await this.save(staging);

          // these are the module dependencies we compile against
          const deps = ['@types/node', 'constructs', 'cdktf'];

          const opts: srcmak.Options = {
            entrypoint: target.fileName,
            deps: deps.map(dep => path.dirname(require.resolve(`${dep}/package.json`))),
            moduleKey: target.moduleKey
          };

          // used for testing.
          if (this.options.outputJsii) {
            opts.jsii = { path: this.options.outputJsii };
          }

          if (this.isPythonTarget) {
            opts.python = {
              outdir: this.codeMakerOutdir,
              moduleName: target.srcMakName
            };
          }

          if (this.isJavaTarget) {
            opts.java = {
              outdir: '.', // generated java files aren't packaged, so just include directly in app
              package: `imports.${target.srcMakName}`
            }
          }

          if (this.isCsharpTarget) {
            opts.csharp = {
              outdir: this.codeMakerOutdir,
              namespace: target.srcMakName
            }
          }

          if (this.isGoTarget) {
            const [orgName , packageName] = target.fqn.split('/');

            opts.golang = {
              // jsii-srcmac will produce a folder inside this dir named after "packageName", so this results in e.g. .gen/hashicorp/random
              outdir: path.join(this.codeMakerOutdir, orgName),
              moduleName: `github.com/terraform-cdk-providers/cdktf-provider-${target.moduleKey.replace(/_/g, '-')}-go`,
              packageName // package will be named e.g. random for hashicorp/random
            }
            console.log({opts, target});
          }

          await srcmak.srcmak(staging, opts);
        });
      }
    }

    for (const target of this.targets) {
      await report(target)
    }
  }

  private async save(outdir = this.codeMakerOutdir) {
    await this.code.save(outdir);
  }

  private get isJavascriptTarget() {
    return this.options.targetLanguage === Language.TYPESCRIPT
  }

  private get isPythonTarget() {
    return this.options.targetLanguage === Language.PYTHON
  }

  private get isJavaTarget() {
    return this.options.targetLanguage === Language.JAVA
  }

  private get isCsharpTarget() {
    return this.options.targetLanguage === Language.CSHARP
  }

  private get isGoTarget() {
    return this.options.targetLanguage === Language.GO
  }
}

const report = async (target: ConstructsMakerTarget): Promise<void> => {
  const reportParams: ReportParams = {
    command: 'get',
    product: 'cdktf',
    version: VERSION,
    dateTime: new Date(),
    payload: target.trackingPayload,
    language: target.targetLanguage
  };
  await ReportRequest(reportParams);
}