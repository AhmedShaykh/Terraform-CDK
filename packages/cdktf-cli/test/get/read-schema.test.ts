import { readSchema } from "../../lib/get/generator/provider-schema";
import { ConstructsMakerModuleTarget, ConstructsMakerProviderTarget, Language } from "../../lib/get/constructs-maker";
import { TerraformModuleConstraint, TerraformProviderConstraint } from "../../lib/config";

describe("readSchema", () => {
  beforeAll(() => {
    jest.setTimeout(60000)
  })

  it("generates a single provider schema", async () => {
    const provider = new TerraformProviderConstraint('hashicorp/null@3.1.0')
    const targets = new ConstructsMakerProviderTarget(provider, Language.TYPESCRIPT)
    const result = await readSchema([targets])
    expect(result).toMatchSnapshot();
  });

  it("generates a single module schema", async () => {
    const module = new TerraformModuleConstraint('terraform-aws-modules/iam/aws//modules/iam-account@3.12.0')
    const targets = new ConstructsMakerModuleTarget(module, Language.TYPESCRIPT)
    const result = await readSchema([targets])
    expect(result).toMatchSnapshot();
  });

  it("generates a more complex schema", async () => {
    const module = new TerraformModuleConstraint('terraform-aws-modules/eks/aws@7.0.1')
    const targets = new ConstructsMakerModuleTarget(module, Language.TYPESCRIPT)
    const result = await readSchema([targets])
    expect(result).toMatchSnapshot();
  });

});
