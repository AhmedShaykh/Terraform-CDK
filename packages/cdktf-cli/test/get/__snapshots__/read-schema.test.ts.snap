// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readSchema generates a single module schema 1`] = `
Object {
  "moduleSchema": Object {
    "iam-account": Object {
      "inputs": Array [
        Object {
          "description": "AWS IAM account alias for this account",
          "name": "account_alias",
          "required": true,
          "type": "string",
        },
        Object {
          "default": true,
          "description": "Whether to allow users to change their own password",
          "name": "allow_users_to_change_password",
          "required": false,
          "type": "bool",
        },
        Object {
          "default": true,
          "description": "Whether to create AWS IAM account password policy",
          "name": "create_account_password_policy",
          "required": false,
          "type": "bool",
        },
        Object {
          "default": true,
          "description": "Whether to get AWS account ID, User ID, and ARN in which Terraform is authorized",
          "name": "get_caller_identity",
          "required": false,
          "type": "bool",
        },
        Object {
          "default": false,
          "description": "Whether users are prevented from setting a new password after their password has expired (i.e. require administrator reset)",
          "name": "hard_expiry",
          "required": false,
          "type": "bool",
        },
        Object {
          "default": 0,
          "description": "The number of days that an user password is valid.",
          "name": "max_password_age",
          "required": false,
          "type": "number",
        },
        Object {
          "default": 8,
          "description": "Minimum length to require for user passwords",
          "name": "minimum_password_length",
          "required": false,
          "type": "number",
        },
        Object {
          "default": null,
          "description": "The number of previous passwords that users are prevented from reusing",
          "name": "password_reuse_prevention",
          "required": false,
          "type": "number",
        },
        Object {
          "default": true,
          "description": "Whether to require lowercase characters for user passwords",
          "name": "require_lowercase_characters",
          "required": false,
          "type": "bool",
        },
        Object {
          "default": true,
          "description": "Whether to require numbers for user passwords",
          "name": "require_numbers",
          "required": false,
          "type": "bool",
        },
        Object {
          "default": true,
          "description": "Whether to require symbols for user passwords",
          "name": "require_symbols",
          "required": false,
          "type": "bool",
        },
        Object {
          "default": true,
          "description": "Whether to require uppercase characters for user passwords",
          "name": "require_uppercase_characters",
          "required": false,
          "type": "bool",
        },
      ],
      "name": "iam-account",
      "outputs": Array [
        Object {
          "description": "The AWS Account ID number of the account that owns or contains the calling entity",
          "name": "this_caller_identity_account_id",
        },
        Object {
          "description": "The AWS ARN associated with the calling entity",
          "name": "this_caller_identity_arn",
        },
        Object {
          "description": "The unique identifier of the calling entity",
          "name": "this_caller_identity_user_id",
        },
        Object {
          "description": "Indicates whether passwords in the account expire. Returns true if max_password_age contains a value greater than 0. Returns false if it is 0 or not present.",
          "name": "this_iam_account_password_policy_expire_passwords",
        },
      ],
    },
  },
  "providerSchema": Object {},
}
`;

exports[`readSchema generates a single provider schema 1`] = `
Object {
  "moduleSchema": Object {},
  "providerSchema": Object {
    "format_version": "0.1",
    "provider_schemas": Object {
      "registry.terraform.io/hashicorp/null": Object {
        "data_source_schemas": Object {
          "null_data_source": Object {
            "block": Object {
              "attributes": Object {
                "has_computed_default": Object {
                  "computed": true,
                  "description": "If set, its literal value will be stored and returned. If not, its value defaults to \`\\"default\\"\`. This argument exists primarily for testing and has little practical use.",
                  "description_kind": "markdown",
                  "optional": true,
                  "type": "string",
                },
                "id": Object {
                  "computed": true,
                  "deprecated": true,
                  "description": "This attribute is only present for some legacy compatibility issues and should not be used. It will be removed in a future version.",
                  "description_kind": "markdown",
                  "type": "string",
                },
                "inputs": Object {
                  "description": "A map of arbitrary strings that is copied into the \`outputs\` attribute, and accessible directly for interpolation.",
                  "description_kind": "markdown",
                  "optional": true,
                  "type": Array [
                    "map",
                    "string",
                  ],
                },
                "outputs": Object {
                  "computed": true,
                  "description": "After the data source is \\"read\\", a copy of the \`inputs\` map.",
                  "description_kind": "markdown",
                  "type": Array [
                    "map",
                    "string",
                  ],
                },
                "random": Object {
                  "computed": true,
                  "description": "A random value. This is primarily for testing and has little practical use; prefer the [hashicorp/random provider](https://registry.terraform.io/providers/hashicorp/random) for more practical random number use-cases.",
                  "description_kind": "markdown",
                  "type": "string",
                },
              },
              "deprecated": true,
              "description": "The \`null_data_source\` data source implements the standard data source lifecycle but does not
interact with any external APIs.

Historically, the \`null_data_source\` was typically used to construct intermediate values to re-use elsewhere in configuration. The
same can now be achieved using [locals](https://www.terraform.io/docs/language/values/locals.html).
",
              "description_kind": "markdown",
            },
            "version": 0,
          },
        },
        "provider": Object {
          "block": Object {
            "description_kind": "plain",
          },
          "version": 0,
        },
        "resource_schemas": Object {
          "null_resource": Object {
            "block": Object {
              "attributes": Object {
                "id": Object {
                  "computed": true,
                  "description": "This is set to a random value at create time.",
                  "description_kind": "markdown",
                  "type": "string",
                },
                "triggers": Object {
                  "description": "A map of arbitrary strings that, when changed, will force the null resource to be replaced, re-running any associated provisioners.",
                  "description_kind": "markdown",
                  "optional": true,
                  "type": Array [
                    "map",
                    "string",
                  ],
                },
              },
              "description": "The \`null_resource\` resource implements the standard resource lifecycle but takes no further action.

The \`triggers\` argument allows specifying an arbitrary set of values that, when changed, will cause the resource to be replaced.",
              "description_kind": "markdown",
            },
            "version": 0,
          },
        },
      },
    },
  },
}
`;
